cheatsheets/Java_Security_Cheat_Sheet.md

```diff
@@ -0,0 +1,685 @@
+
+```java
+import java.nio.charset.StandardCharsets;
+import java.security.SecureRandom;
+import javax.crypto.spec.*;
+import javax.crypto.*;
+import java.util.Base64;
+
+
+// AesGcmSimpleTest
+class Main {
+
+    public static void main(String[] args) throws Exception {
+        // Key of 32 bytes / 256 bits for AES
+        KeyGenerator keyGen = KeyGenerator.getInstance(AesGcmSimple.ALGORITHM);
+        keyGen.init(AesGcmSimple.KEY_SIZE, new SecureRandom());
+        SecretKey secretKey = keyGen.generateKey();
+
+        // Nonce of 12 bytes / 96 bits and this size should always be used.
+        byte[] nonce = new byte[AesGcmSimple.IV_LENGTH];
+        SecureRandom random = new SecureRandom();
+        random.nextBytes(nonce);
+
+        var message = "This message to be encrypted";
+        System.out.println(message);
+
+        // Encrypt the message
+        byte[] cipherText = AesGcmSimple.Encrypt(message, nonce, secretKey);
+        System.out.println(Base64.getEncoder().encodeToString(cipherText));
+
+        // Decrypt the message
+        var message2 = AesGcmSimple.Decrypt(cipherText, nonce, secretKey);
+        System.out.println(message2);
+    }
+}
+
+class AesGcmSimple {
+
+    public static final String ALGORITHM = "AES";
+    public static final String CIPHER_ALGORITHM = "AES/GCM/NoPadding";
+    public static final int KEY_SIZE = 256;
+    public static final int TAG_LENGTH = 128;
+    public static final int IV_LENGTH = 12;
+
+    public static byte[] Encrypt(String plaintext, byte[] nonce, SecretKey secretKey) throws Exception {
+        return CryptoOperation(plaintext.getBytes(StandardCharsets.UTF_8), nonce, secretKey, Cipher.ENCRYPT_MODE);
+    }
+
+    public static String Decrypt(byte[] ciphertext, byte[] nonce, SecretKey secretKey) throws Exception {
+        return new String(CryptoOperation(ciphertext, nonce, secretKey, Cipher.DECRYPT_MODE), StandardCharsets.UTF_8);
+    }
+
+    private static byte[] CryptoOperation(byte[] text, byte[] nonce, SecretKey secretKey, int mode) throws Exception {
+        Cipher cipher = Cipher.getInstance(CIPHER_ALGORITHM);
+        GCMParameterSpec gcmParameterSpec = new GCMParameterSpec(TAG_LENGTH, nonce);
+        cipher.init(mode, secretKey, gcmParameterSpec);
+        return cipher.doFinal(text);
+    }
+
+}
+```
+
+### Encryption for transmission
+
+Again, follow the algorithm guidance in the [OWASP Cryptographic Storage Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/Cryptographic_Storage_Cheat_Sheet.html#algorithms).
+
+The following code snippet shows an example of using Eliptic Curve/Diffie Helman (ECDH) together with AES-GCM to perform encryption/decryption of data between two different sides without the need the transfer the symmetric key between the two sides. Instead, the sides exchange public keys and can then use ECDH to generate a shared secret which can be used for the symmetric encryption.

```

Comment:
No, no, no. A thousand times NO. There is much more to it than this and you are essentially ignoring all the lessons learned in TLS v1.3.  If you want to illustrate this, don't show it from first principles, but instead show them how to configure TLS v1.3 and use that in Java. All the cipher suites in TLS v1.3 are thought to be safe. This, OTOH, is not, as it does not ensure that you ensure the authenticity of the intended end-point.

Reply:
I hear that. The concept was to try and provide a solution for a scenario where we want to encrypt something in addition to TLS (e.g. a particularly sensitive data item). I have seen this done before using AES to encrypt the data, then encryption the AES DEK with an RSA KEK and sending the encrypted DEK and the encrypted data to the other side. I was hoping to try this using eliptic curve and I ended up with this mechanism because ECIES is not supported out of the box in Java.

What solution would you suggest for this use case?

Reply:
So, do you mean a hybrid solution? Generally you only do that when you have many parties who want to encrypt data and share the encrypted value with a common recipient, but also keep it from each other. In such cases, using a shared key won't work because other parties that intercept it and share that secret key could decrypt it. To the sender creates a temporary  symmetric ephemeral key (what you are referring to as your 'AES DEK'), and they encrypt their sensitive data using symmetric encryption with that. They then encrypt that symmetric ephemeral key with the public asymmetric key of the recipient and typically prepend that to the beginning of the symmetric ciphertext.

Yes, that's a legitimate use case that's useful because it greatly simplifies the key distribution problem. But that didn't look like what you were doing there. (Or maybe it was, but I was just getting tired at that point.)

Regardless, there's a hybrid example for Google Tink here: https://github.com/tink-crypto/tink-java/tree/main/examples/hybrid

Reply:
I haven't checked, but Tink likely supports [Cryptographic Message Syntax (CMS)](https://en.wikipedia.org/wiki/Cryptographic_Message_Syntax). If .NET also support that (which wouldn't surprise me, that's what I would suggest as that's commonly used in a lot of standards. If that is not available, export it all as JSON and have .NET parse it. That's how I would address the interoperability issue. Way better than writing something that's likely to be error prone and that I don't presently have time for a line-by-line detailed code review.